python/activator/activator.py

```diff
@@ -257,73 +324,95 @@ def keda_start():
 
-            fan_out_message = fan_out_consumer.poll(timeout=5)
-            if fan_out_message is None:
-                continue
-            if fan_out_message.error():
-                _log.warning("Fanned out consumer error: %s", fan_out_message.error())
-            else:
-                deserialized_fan_out_visit = fan_out_avro_deserializer(fan_out_message.value(),
-                                                                       SerializationContext(
-                                                                       fan_out_message.topic(),
-                                                                       MessageField.VALUE))
-                _log.info("Unpacked message as %r.", deserialized_fan_out_visit)
-
-                # Calculate time to load knative and receive message based on timestamp in Kafka message
-                _log.debug("Message timestamp %r", fan_out_message.timestamp())
-                fan_out_kafka_msg_timestamp = fan_out_message.timestamp()
-                fan_out_to_prompt_time = int(time.time() * 1000) - fan_out_kafka_msg_timestamp[1]
-                _log.debug("Seconds since fan out message delivered %r", fan_out_to_prompt_time/1000)
-
-                # Commit message and close client
-                fan_out_consumer.commit(message=fan_out_message, asynchronous=False)
-                fan_out_consumer.close()
+                try:
+                    fan_out_message = redis_client.xreadgroup(
+                        streams={redis_stream_name: ">"},
+                        consumername=redis_group_id,
+                        groupname=redis_stream_consumer_group,
+                        count=1,)
 
+                    if not fan_out_message:
+                        continue
+                    else:
+                        consumer_polls_with_message += 1
+                        if consumer_polls_with_message >= 1:
+                            fan_out_listen_time = _calculate_time_since_last_message(
+                                fan_out_listen_start_time)
+                            _log.debug(
+                                "Seconds since last redis streams message received %r for consumer poll %r",
+                                fan_out_listen_time, consumer_polls_with_message)
+
+                        redis_streams_message_id, fan_out_visit_decoded = _decode_redis_streams_message(
+                            fan_out_message)
+
+                        # Calculate time to receive message based on timestamp in Redis Stream message
+                        fan_out_to_prompt_time = _calculate_time_since_fan_out_message_delivered(
+                            redis_streams_message_id)
+                        _log.debug("Seconds since fan out message delivered %r", fan_out_to_prompt_time)
+
+                        expected_visit = FannedOutVisit.from_dict(fan_out_visit_decoded)
+                        _log.debug("Unpacked message as %r.", expected_visit)
```

Comment:
I suggest moving this code to the same `try` block as `process_visit`. This will ensure that a) `process_visit` is only called if `expected_visit` is defined, and b) and in parsing a single message don't break out of the loop.

You might need to switch the order of `instances_processing_gauge.track_in_progress` if message parsing isn't supposed to be included.